File: python/day22.py
from aoc import *
import re


pattern = re.compile(r"(L|R)")


def field2grid(field):
    return {x+y*1j: c
            for y, line in enumerate(field)
            for x, c in enumerate(line)
            if c in "#."}


def wrap_1(pt, d):
    """Hardcoded for my input's shape
          A B
          C
        D E
        F
    """
    x = pt.real
    y = pt.imag * 1j
    if d in {1, -1}:
        match pt.imag//50:
            case 0 | 2: return (x - 100*d) + y, d
            case 1 | 3: return (x -  50*d) + y, d
    else:
        match x//50:
            case 0: return x + (y - 100*d), d
            case 1: return x + (y - 150*d), d
            case 2: return x + (y -  50*d), d


def wrap_2(pt, d):
    """Hardcoded for my input's shape
          A B
          C
        D E
        F
    """
    x, y = pt.real, pt.imag
    match x//50, y//50, d:
        case _, 0, -1:  return complex(0, 149-y), 1     # A --> D
        case _, 1, -1:  return complex(y-50, 100), 1j   # C --> D
        case _, 2, -1:  return complex(50, 149-y), 1    # D --> A
        case _, 3, -1:  return complex(y-100, 0), 1j    # F --> A
        case _, 0,  1:  return complex(99, 149-y), -1   # B --> E
        case _, 1,  1:  return complex(y+50, 49), -1j   # C --> B
        case _, 2,  1:  return complex(149, 149-y), -1  # E --> B
        case _, 3,  1:  return complex(y-100, 149), -1j # F --> E
        case 0, _, -1j: return complex(50, x+50), 1     # D --> C
        case 1, _, -1j: return complex(0, x+100), 1     # A --> F
        case 2, _, -1j: return complex(x-100, 199), -1j # B --> F
        case 0, _,  1j: return complex(x+100, 0), 1j    # F --> B
        case 1, _,  1j: return complex(49, x+100), -1   # E --> F
        case 2, _,  1j: return complex(99, x-50), -1    # B --> C


def traverse(pos, moves, grid, wrap):
    d = 1
    for move in moves:
        match move:
            case 'L': d *= -1j
            case 'R': d *=  1j
            case _:
                for _ in range(int(move)):
                    new_pos = pos + d
                    d_ = d
                    if new_pos not in grid:
                        new_pos, d_ = wrap(new_pos, d)
                    if grid[new_pos] == '.':
                        pos = new_pos
                        d = d_
                    else:
                        break
    return pos, d


def password(pos, d):
    x = pos.real + 1
    y = pos.imag + 1
    facing = {1: 0, 1j: 1, -1: 2, -1j: 3}[d]
    return int(4*x + 1000*y + facing)


def solve(filename=22):
    *field, _, path = read_input(filename)
    start = field[0].find('.')
    moves = re.split(pattern, path)
    grid = field2grid(field)
    p1 = traverse(start, moves, grid, wrap_1)
    p2 = traverse(start, moves, grid, wrap_2)
    return password(*p1), password(*p2)


print(solve())










def test():
    assert solve() == (13566, 11451)

File: clojure/day22.clj
(ns day22
  (:require aoc
            [clojure.string :as str]
            [clojure.core.match :refer [match]]))


(defn wrap-1 [[x y] [dx dy]]
  ;; hardcoded for my input
  (cond
    (zero? dy) (let [wrap (case (quot y 50) , (0 2) 100 , (1 3) 50)]
                 [[(- x (* wrap dx)) y] [dx dy]])
    (zero? dx) (let [wrap (case (quot x 50) , 0 100 , 1 150 , 2 50)]
                 [[x (- y (* wrap dy))] [dx dy]])))

(defn wrap-2 [[x y] d]
  ;; hardcoded for my input
  (match [(quot x 50) (quot y 50) d]
    [_ 0 [-1 0]] [[0 (- 149 y)]   [1  0]]
    [_ 1 [-1 0]] [[(- y 50) 100]  [0  1]]
    [_ 2 [-1 0]] [[50 (- 149 y)]  [1  0]]
    [_ 3 [-1 0]] [[(- y 100) 0]   [0  1]]
    [_ 0 [1  0]] [[99 (- 149 y)]  [-1 0]]
    [_ 1 [1  0]] [[(+ 50 y) 49]   [0 -1]]
    [_ 2 [1  0]] [[149 (- 149 y)] [-1 0]]
    [_ 3 [1  0]] [[(- y 100) 149] [0 -1]]
    [0 _ [0 -1]] [[50 (+ 50 x)]   [1  0]]
    [1 _ [0 -1]] [[0 (+ 100 x)]   [1  0]]
    [2 _ [0 -1]] [[(- x 100) 199] [0 -1]]
    [0 _ [0  1]] [[(+ x 100) 0]   [0  1]]
    [1 _ [0  1]] [[49 (+ 100 x)]  [-1 0]]
    [2 _ [0  1]] [[99 (- x 50)]   [-1 0]]))



(defn walk [grid wrap-fn move pos d]
  (reduce
   (fn [[p d] _]
     (let [p'       (aoc/pt+ p d)
           [p'' d'] (if-not (contains? grid p')
                      (wrap-fn p' d)
                      [p' d])]
       (if (= \. (grid p'' \#))
         [p'' d']
         (reduced [p d]))))
   [pos d]
   (range (parse-long move))))


(defn traverse [start moves grid wrap-fn]
  (let [walk_ (partial walk grid wrap-fn)]
    (reduce
     (fn [[pos [dx dy :as d]] move]
       (case move
         "L" [pos [dy (- dx)]]
         "R" [pos [(- dy) dx]]
         (walk_ move pos d)))
     [start [1 0]]
     moves)))

(defn password [[[x y] d]]
  (+ (* 4 (inc x))
     (* 1000 (inc y))
     ({[1 0] 0 , [0 1] 1 , [-1 0] 2 , [0 -1] 3} d)))

(defn solve
  ([] (solve (aoc/read-file 22)))
  ([input]
   (let [[field [moves]] (aoc/parse-input-paragraphs input)
         start  [(str/index-of (first field) \.) 0]
         moves  (-> moves
                    (str/replace #"(L|R)" " $1 ")
                    (str/split #" "))
         grid   (aoc/grid->points field #{\. \#})
         move_  (partial traverse start moves grid)]
     [(-> (move_ wrap-1) password)
      (-> (move_ wrap-2) password)])))


(solve)
